### ISSUE TITLE
expo-auth-session does not work in ejected app

### ISSUE THREAD

**Issue by celandro** [expo/expo#7816]
## 🐛 Bug Report

In general the documentation is missing or incomplete for both Linking and for AuthSession for bare apps and its hours of googling and trial and error to try and get things working.  I managed to get IOS working but Android is still perplexing me. A blog post would be a huge help.

Issue with IOS is missing URL identifier in url types. Expo-Auth-Session did not work at all before
Expected
```
<key>CFBundleURLTypes</key>
        <array>
                <dict>
                        <key>CFBundleTypeRole</key>
                        <string>Editor</string>
                        <key>CFBundleURLName</key>
                        <string>pb</string>
                        <key>CFBundleURLSchemes</key>
                        <array>
                                <string>pb</string>
                        </array>
                </dict>
        </array>
```
Actual
```
<key>CFBundleURLTypes</key>
        <array>
                <dict>
                        <key>CFBundleTypeRole</key>
                        <string>Editor</string>
                        <key>CFBundleURLSchemes</key>
                        <array>
                                <string>pb</string>
                        </array>
                </dict>
        </array>
```

Android is missing intents AND a custom URL handler.
Example intent:
```
            <intent-filter>
                <action android:name="android.intent.action.VIEW"/>
                <data android:scheme="pb"/>
                <category android:name="android.intent.category.BROWSABLE"/>
                <category android:name="android.intent.category.DEFAULT"/>
            </intent-filter>
```

### Environment

  Expo CLI 3.17.23 environment info:
    System:
      OS: macOS 10.15.4
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 13.12.0 - /usr/local/bin/node
      npm: 6.14.4 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    IDEs:
      Android Studio: 3.6 AI-192.7142.36.36.6308749
      Xcode: 11.4/11E146 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^37.0.0 => 37.0.7 
      react: ~16.9.0 => 16.9.0 
      react-native: ~0.61.5 => 0.61.5 
    npmGlobalPackages:
      expo-cli: 3.17.23

### Steps to Reproduce

1. Create an expo app with expo-auth-session and a custom scheme
1. expo eject
1. Modify code to set custom redirectUrl and return url
1. Modify 3rd party oauths to accept new return url

### Expected Behavior
Login works

### Actual Behavior

Login fails on final load of pb://raidparty/login/success

* IOS fix is above.

* Android fix is in comment below

**Comment by byCedric** [comment]
Thanks for splitting this out from #7807! @brentvatne there is also a known issue with the manifest properties with auth-session in the bare flow, you might run into it but should be fixed already (see #7807)

**Comment by celandro** [comment]
With the intent, the login is actually coming back as cancelled and then crashing due to a separate issue in native base toast. Still looking into it on my side

**Comment by celandro** [comment]
Some more detail. It is actually coming back as 'dismiss'. I looked at similar related bug #6679  and it does not seem to be the same issue. I applied the patch in the comments and it does not resolve the issue.

**Comment by celandro** [comment]
I think I figured it out.  The url listener is never being called
I added some logs

```
function _waitForRedirectAsync(returnUrl) {
    return new Promise(resolve => {
        _redirectHandler = (event) => {
            console.log('event.url:', event.url)
            if (event.url.startsWith(returnUrl)) {
                resolve({ url: event.url, type: 'success' });
            }
        };
        console.log('Adding redirect handler')
        Linking.addEventListener('url', _redirectHandler);
    });
}
```

And the `event.url:'` log line is never showing up

**Comment by celandro** [comment]
Can confirm missing scheme in  Android activity should either be the intent as above or something like:

```
            <intent-filter>
                <action android:name="android.intent.action.VIEW"/>
                <category android:name="android.intent.category.BROWSABLE"/>
                <category android:name="android.intent.category.DEFAULT"/>
                <data android:scheme="pb" android:host="raidParty" />
            </intent-filter>
```

Preferably this should be done in eject but documentation and a TODO (similar to icons/splash screen) would be sufficient

**Comment by celandro** [comment]
https://github.com/droibit/react-native-custom-tabs/issues/27 implies that url listener for custom tabs is not implemented intentionally.

I assume the correct thing to do is to make a route that maps to the successful end url and allow Linking to work?

**Comment by celandro** [comment]
Can confirm, to make this work on ios, you must add a deep link handler for your custom url. I am using react-navigation and followed https://reactnavigation.org/docs/deep-linking/

Added:
```
function LoginHandler({ route }) {
    let {params={}} = route;
    let {jwt, status} = params;
    return (
        <View style={{flex: 1, justifyContent: "center", alignItems: "center"}}>
            <Text>Login {status? status: "Unknown Status"}</Text>
            <Text>
                JWT: {jwt ? jwt : "No jwt found"}
            </Text>
        </View>
    );
}
...
            <Stack.Screen name="LoginHandler" component={LoginHandler} />
...
const config = {
    LoginHandler: {
        path: "raidParty/login/:status",
        parse: {
            status: String
        },
    }
};

function Nav() {
    const ref = React.useRef();
    const { getInitialState } = useLinking(ref, {
        prefixes: ['pb://'],
        config
    });
    const [isReady, setIsReady] = React.useState(false);
    const [initialState, setInitialState] = React.useState();

    React.useEffect(() => {
        Promise.race([
            getInitialState(),
            new Promise(resolve =>
                // Timeout in 150ms if `getInitialState` doesn't resolve
                // Workaround for https://github.com/facebook/react-native/issues/25675
                setTimeout(resolve, 150)
            ),
        ])
            .catch(e => {
                console.error(e);
            })
            .then(state => {
                if (state !== undefined) {
                    setInitialState(state);
                }

                setIsReady(true);
            });
    }, [getInitialState]);

    if (!isReady) {
        console.log('No deep link found')
        return null;
    }
    console.log(`Initial state: ${JSON.stringify(initialState)}`);

    return (<NavigationContainer initialState={initialState} ref={ref}>
            <AuthStack/>
        </NavigationContainer>
    )
}

export default Nav
```

For IOS, the url listening works as b but now there are 2 code paths to test which is undesirable.

So in summary I'm unblocked but look forward to a more unified implementation. I hope this issue helps anyone who tries to use expo-auth-session in a bare app.

**Comment by EvanBacon** [comment]
The documentation problem could partially be resolved by the new proposed docs changes in #7760 https://github.com/expo/expo/blob/@evanbacon/auth-session/auth-request/docs/pages/versions/unversioned/sdk/auth-session.md#installation

**Comment by celandro** [comment]
That would have definitely made the process much smoother and handles the first 2 problems.

The third issue with needing to add a linking handler on android due to moving to androidX needs to be more explicit for bare apps.  Totally separately I have no idea how auth  session works in a non bare app due to the url listener issue

**Comment by celandro** [comment]
Final comment for now, I tried the documentation from the link. IOS worker properly

Did not work:
```
android.defaultConfig.manifestPlaceholders = [
  'appAuthRedirectScheme': 'com.pokebattler.raidparty'
]
```


```
<activity
        android:name="net.openid.appauth.RedirectUriReceiverActivity"
        tools:node="replace">
    <intent-filter>
        <action android:name="android.intent.action.VIEW"/>
        <category android:name="android.intent.category.DEFAULT"/>
        <category android:name="android.intent.category.BROWSABLE"/>
        <data android:scheme="com.pokebattler.raidparty"/>
    </intent-filter>
</activity>
```
crashed the simulator.

Adding the intent to the MainActivity AND adding a listener does work

**Comment by EvanBacon** [comment]
The new AuthSession API is ready for use https://docs.expo.io/versions/latest/sdk/auth-session/ please see if using `uri-scheme` helps to debug the issue. I've tested that new auth-session works in `apps/bare-expo` on bare workflow for Android and iOS across 14 different providers. I suspect that there might be an issue with your `AndroidManifest.xml`, please refer to this troubleshooting guide I put together: https://expo.fyi/setup-android-uri-scheme

**Comment by hbarylskyi** [comment]
troubleshooting guide helped me, thanks @EvanBacon
